include/zephyr/net/socket_service.h

```diff
@@ -0,0 +1,180 @@
+		.work_q = (_work_q),                                    \
+		.pev = __z_net_socket_svc_get_name(_name),		\
+		.pev_len = (_count),					\
+	}
+
+/**
+ * @brief Statically define a network socket service.
+ *
+ * The socket service can be accessed outside the module where it is defined using:
+ *
+ * @code extern struct net_socket_service <name>; @endcode
+ *
+ * @note This macro cannot be used together with a static keyword.
+ *       If such a use-case is desired, use @ref NET_SOCKET_SERVICE_DEFINE_STATIC
+ *       instead.
+ *
+ * @param name Name of the service.
+ * @param work_q Pointer to workqueue where the work is done. Can be null in which case
+ *        system workqueue is used.
+ * @param cb Callback function that is called for socket activity.
+ * @param count How many pollable sockets is needed for this service.
+ */
+#define NET_SOCKET_SERVICE_DEFINE(name, work_q, cb, count)	\
+	__z_net_socket_service_define(name, work_q, cb, count)
+
+/**
+ * @brief Statically define a network socket service in a private (static) scope.
+ *
+ * @param name Name of the service.
+ * @param work_q Pointer to workqueue where the work is done. Can be null in which case
+ *        system workqueue is used.
+ * @param cb Callback function that is called for socket activity.
+ * @param count How many pollable sockets is needed for this service.
+ */
+#define NET_SOCKET_SERVICE_DEFINE_STATIC(name, work_q, cb, count)	\
+	__z_net_socket_service_define(name, work_q, cb, count, static)
+	
+
+/**
+ * @brief Register pollable sockets.
+ *
+ * @param service Pointer to a service description.
+ * @param fds Socket array to poll.
+ * @param len Lenght of the socket array.
+ * @param user_data User specific data.
+ *
+ * @retval 0 No error
+ * @retval -ENOENT Service is not found.
+ * @retval -ENINVAL Invalid parameter.
+ */
+__syscall int net_socket_service_register(struct net_socket_service_desc *service,
+					  struct zsock_pollfd *fds, int len, void *user_data);
+
+/**
+ * @brief Unregister pollable sockets.
+ *
+ * @param service Pointer to a service description.
+ *
+ * @retval 0 No error
+ * @retval -ENOENT Service is not found.
+ * @retval -ENINVAL Invalid parameter.
+ */
+static inline int net_socket_service_unregister(struct net_socket_service_desc *service)
+{
+	return net_socket_service_register(service, NULL, 0, NULL);
+}
+
+/**
+ * @typedef net_socket_service_cb_t
+ * @brief Callback used while iterating over socket services.
+ *
+ * @param svc Pointer to current socket service.
+ * @param user_data A valid pointer to user data or NULL
+ */
+typedef void (*net_socket_service_cb_t)(struct net_socket_service_desc *svc,
+					void *user_data);
+
+/**
+ * @brief Go through all the socket services and call callback for each interface.
+ *
+ * @param cb User-supplied callback function to call
+ * @param user_data User specified data
+ */
+void net_socket_service_foreach(net_socket_service_cb_t cb, void *user_data);
```

Comment:
I wonder, what's intended use case for such functionality?

Reply:
It was for the net-shell. I was planning to have some monitoring information provided by shell and there this function would be useful. I did not implemented the shell part yet as so many things are still open with the APIs etc.